### DISCUSSION TITLE
High web server load due to "chocolatey command line" clients

### DISCUSSION THREAD

**Discussion by boardhead** [chocolatey/home#299] [Q&A]
My web site shows intermittent periods of high server load due to various IP's using "chocolatey command line" and downloading multiple copies of my software (ExifTool). Often 70 MB or more is transferred for a package that is only 7 MB in size. Why would chocolatey-command-line clients be downloading a distribution package multiple times? I've been banning these IP's to keep server load under control.

Any ideas what is going on? Any way to mitigate this?

- Phil

**Comment by pauby** [comment]
Downloading 70MB for a 7MB file doesn't make sense. I'm assuming the package is [here](https://community.chocolatey.org/packages/exiftool)? As the license permits distribution, it may be worthwhile embedding the downloaded file inside the package so that your website is only used once to download it to put into the package. @mkevenaar is this something that could be done?

Are you able to ping me on Discord as @pauby ? We can then grab some logs from you and may be able to help better understand what is going on.

**Reply by mkevenaar** [reply]
Sure, I can alter the package to be an embedded package. Please create an issue for that [here](https://github.com/mkevenaar/chocolatey-packages/issues)

**Reply by pauby** [reply]
[Done](https://github.com/mkevenaar/chocolatey-packages/issues/230).

**Reply by pauby** [reply]
I noticed on your website this comment:

> (@chocolatey: Clean up your act because your clients will get banned when they download the distribution package 20 times!)

We have no control over who is using the Chocolatey client to install software from packages, and we don't have any control over how often they do that. If you feel you need to ban IP's for excessive downloading, we have no issue with that. We have excessive use policies that cover the Chocolatey Community Repository. Reaching out earlier, so we can help, would have been a better option. If we are not aware of issues, we can't be expected to do anything about them.

**Reply by gep13** [reply]
@boardhead on a side note... are you using something like Cloudflare in front of your site, so that you don't need to worry about how many downloads there are?

Happy to chat with you about this if this isn't something you have in place.

**Reply by boardhead** [reply]
Thanks for your responses.   Embedding the binary into that package should solve the issue if that package is responsible for these downloads.

@gep13 I'm not using CloudFlare on my site.  Email me directly and we can talk about this (philharvey66 at gmail.com).

**Reply by mkevenaar** [reply]
@boardhead a new version of ExifTool has been pushed to Chocolatey: https://community.chocolatey.org/packages/exiftool/12.76.0.20240210 

This version should go live automagically once all the automation is completed. This should reduce the amount of requests of ExifTool significantly. This can be up to 10k per version, based on the Chocolatey package downloads. 

If you do track the amount of downloads per version, I would advise you to include the numbers shown here: https://community.chocolatey.org/packages/exiftool/#versionhistory

**Reply by boardhead** [reply]
Great, thanks!  I'll let you know if I see any more problems.

There are so many ExifTool mirrors out there that I couldn't possibly hope keep track of the number of downloads.

- Phil

**Reply by boardhead** [reply]
I don't know if the solution was supposed to be immediate, but here is an example from 3 hours ago where a chocolatey client downloaded 127 MB:

```
91.xx.xx.xx - - [11/Feb/2024:01:00:13 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:13 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:13 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:13 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:14 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:14 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:14 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:14 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:14 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:15 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:15 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:15 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:15 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:15 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:16 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:21 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:21 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:22 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 119253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 71653 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 77253 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:23 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:24 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:25 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:25 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:25 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line" 
91.xx.xx.xx - - [11/Feb/2024:01:00:25 -0800] "GET /exiftool-12.76.zip HTTP/1.1" 200 7319377 "-" "chocolatey command line"
```

- Phil

**Reply by pauby** [reply]
I think it's important to use the correct syntax here as it's becoming confusing.

Those logs don't indicate 'a Chocolatey client'. Those logs indicate an IP address. Behind that IP address is multiple Chocolatey clients. So multiple Chocolatey clients are downloading exiftool. One Chocolatey client is not downloading 127MB.

The latest version, version 12.76.0.20240210, [includes the `exiftool.zip` inside the package](https://community.chocolatey.org/packages/exiftool#files). Previous versions do not. If those downloads are installing 12.76.0 of the package, they will download from your website.

**Reply by boardhead** [reply]
I was talking about a client for my web server, but I accept your definition.  I guess I don't understand the chocolatey client model.  It sounds like you are saying these downloads are going to multiple different users.  Why do they come in bursts like this?  And why can't the data be cached for the multiple users if all the downloads are going through the same IP?

- Phil

**Reply by pauby** [reply]
> I guess I don't understand the chocolatey client model.

Chocolatey CLI is a package manager. If you're familiar with package manager’s then hopefully that helps to make sense of what its purpose is. If not, please let me know and I will explain a bit more.

In terms of `exiftool.zip` it makes it easier for people to install and upgrade your software. They don't have to download it from the website manually each time, and install it. The package does that for them.

> It sounds like you are saying these downloads are going to multiple different users. 

Yes. I would _guess_ it's an organization (based on the number of simultaneous downloads). Chocolatey CLI only runs when the user tells it to. It doesn't run in the background, downloading packages and their installers as it wants. So a user has to have physically run the command `choco install exiftool --version 12.76.0`, OR, if they already have the package installed, `choco upgrade exiftool --version 12.76.0`.  I am adding the version there because running without the version would install the latest version that has the `exiftool.zip` file inside the package and therefore doesn't require it to be downloaded from your website. 

In terms of the downloads, it would be like _each_ of those users went to your website and manually downloaded the `exiftool.zip` file.

> Why do they come in bursts like this? 

I'm guessing again, but it would likely be a scheduled task that runs at 01:00 (according to the logs you provided). Chocolatey CLI does not create a scheduled task, so this would be something that the users of those computers have done specifically.

> And why can't the data be cached for the multiple users if all the downloads are going through the same IP?

I'm unsure who this is aimed at.

We don't cache your `exiftool.zip` file for open-source users because it's not hosted here. The file needs to be cached where it's hosted, so that would be something you would need to do.

If you are unhappy the same IP address is downloading the `exiftool.zip` file at 01:00 each day, block the IP address. We do not condone organizations operating this way and recommend they [set up an internal repository](https://docs.chocolatey.org/en-us/guides/organizations/organizational-deployment-guide) and keep the files, with the installers, there. Then, when the computers in their organization want to install or upgrade software, they use that repository and not the public one.

We have a disclaimer on the site when you go to it for the first time:

![image](https://github.com/chocolatey/home/assets/12760779/486c3920-f7ff-4f24-9189-55285f9a3d7d)

The Learn more button at the bottom goes to our [docs](https://docs.chocolatey.org/en-us/community-repository/community-packages-disclaimer).

**Reply by boardhead** [reply]
Thanks.  That explains a lot.  I think I understand things a lot better now.

- Phil

**Comment by boardhead** [comment]
@pauby said:

> Reaching out earlier, so we can help, would have been a better option.

Yes of course.  It took me a while to figure out where to reach out.  I've taken down that message now since it looks like we're making good progress.

- Phil
